### PR TITLE
Refactor MemoryScopes to be self-contained 

### PR DESCRIPTION
**Pull request by tomlm**
Fixes https://github.com/microsoft/botbuilder-dotnet/issues/2842
MemoryScopes were being initialized by external DM, which is not optimal
* it makes it harder to add new scopes
* it adds a hard dependency on the implementation of User and Converation scopes
This delta adds unit test for the issue, but implements BotStateMemoryScope with User and Conversation versions of it which are self contained.
User and Conversation state were being copied into the TurnState, but there was no real reason to do this.
Turn memory scope was always being intialized even if it was already there.

This cleanup allows for
* more BotState based scopes to be easily added (such as Tenant, etc)
* Better encapsulation
* Ability for customers to replace memory scopes (such as User and Conversation) with their own implementations.
To test this I needed to update the DialogStateMemory tests to properly create a DialogContext for executing within, as the UserState/ConversationState memory scopes need actualy UserState and ConversationState accessible off of the turn context.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContainer.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ protected DialogContainer(string dialogId = null)
         {
         }
 
-        protected DialogSet Dialogs { get; set; } = new DialogSet();
+        public DialogSet Dialogs { get; set; } = new DialogSet();
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogContainer.cs]: >public [](start = 8, length = 6)

AdaptiveDialog had public property which is an fxcop error, 

This should be public

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogManager.cs [hunk 2]
```diff
@@ -98,13 +94,9 @@ public async Task<DialogManagerResult> OnTurnAsync(ITurnContext context, Cancell
...
 
             // get dialog stack 
+            var dialogsProperty = conversationState.CreateProperty<DialogState>(DIALOGS);
             DialogState dialogState = await dialogsProperty.GetAsync(context, () => new DialogState(), cancellationToken: cancellationToken).ConfigureAwait(false);
 
-            var namedScopes = MemoryScope.GetScopesMemory(context);
-            namedScopes[ScopePath.USER] = await userScope.GetAsync(context, () => new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase), cancellationToken: cancellationToken).ConfigureAwait(false);
-            namedScopes[ScopePath.CONVERSATION] = await conversationScope.GetAsync(context, () => new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase), cancellationToken: cancellationToken).ConfigureAwait(false);
-            namedScopes[ScopePath.TURN] = new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase);
-
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogManager.cs]: this didn't make any sense. It added a hard dependency on how user and conversation state get initialized.  I moved this into BotStateMemoryScope and we are completely free to change the implementation of what resolves "user" => memory or "Conversation" => memory or any other scope for that matter

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogManager.cs (hunk 2 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogManager.cs]: Oh that makes sense

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.5.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.5.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.5.3)

**Pull request merged**